XMonad/Layout/DecorationEx.hs

```diff
@@ -0,0 +1,102 @@
+
+-----------------------------------------------------------------------------
+-- |
+-- Module      :  XMonad.Layout.DecorationEx
+-- Description :  Advanced window decorations module for XMonad
+-- Copyright   :  (c) 2007 Andrea Rossato, 2009 Jan Vornberger, 2023 Ilya Portnov
+-- License     :  BSD-style (see xmonad/LICENSE)
+--
+-- Maintainer  :  portnov84@rambler.ru
+-- Stability   :  unstable
+-- Portability :  unportable
+--
+-- This set of modules contains a set of type classes and their implementations
+-- which define a flexible and extensible mechanism of window decorations.
+--
+-- Within this mechanism, there are the following entities which define
+-- how decorations will look and work:
+--
+-- * Main object is @DecorationEx@ layout modifier. It is from where everything
+--   starts. It creates, shows and hides decoration windows (rectangles) when
+--   needed. It is parameterized with decoration geometry, decoration engine and
+--   theme. It calls these components to do their parts of the work.
+-- * @DecorationGeometry@ defines where decoration rectangles should be placed.
+--   For example, standard horizontal bar above each window; or tab bar.
+-- * @DecorationEngine@ defines how decorations look and how they react on clicks.
+--   Different implementations of the decoration engine can use different APIs
+--   to draw decorations. Within this package, there is one implementation 
+--   (@TextDecoration@), which uses plain Xlib calls, and displays decoration
+--   widgets with text fragments, like @[X]@ or @[_]@. Other engines can, for
+--   example, use the Cairo library to draw nice gradients and image-based widgets.
+-- * A Decoration widget is an element placed on a window decoration. It defines how
+--   it looks and how it responds to clicks. Examples include usual window 
+--   buttons (minimize, maximize, close), window icon, window title.
+-- * A Decoration theme defines colors and fonts for the decoration engine. It also
+--   contains a list of decoration widgets and says where to place them (at the
+--   left, at the right or in the center).
+-- 
+-- This mechanism makes major use of parameterized data types and type families,
+-- in order to make it possible to define different types of decorations, and
+-- easily combine different aspects of decorations. For example, each decoration
+-- engine can be combined with each decoration geometry.
+-----------------------------------------------------------------------------
```

Comment:
I think it would help users a lot to add some of your screenshots here (much like we did [here](https://github.com/xmonad/xmonad-contrib/pull/848/files#diff-021f53dc93fd53911c56f2a644d91d3744b45db5266690f188ab601deee5bc41R47))

Reply:
I'm not sure how to do it properly. I used to just add an image in github comments and then give a link to it. But images in current thread point to private-user-images.githubusercontent.com, and when I click on them, sometimes they open normally, sometimes just blank screen opens. I googled that problem, but still don't quite understand. Either that is just a glitch of github, or it refuses to work as image hosting... Probably it has something to do with my permissions on this repository. But then, most people who will read documentation will not have any specific privileges either.

Reply:
Mh, I did notice the `private-user-images.githubusercontent.com` as well with your images. Perhaps if you just send them my way somehow, I can—presumably—create longer lasting links (at least the one in #848 uses a `github.com/xmonad/xmonad-contrib/assets` URL)

Reply:
https://ibb.co/k6k1Nsf this is the one from description of this PR.